test/unit/test_keyring_multi.py

```diff
@@ -0,0 +1,242 @@
+    assert exc_info.match(
+        "Generator keyring not provided and encryption materials do not already " "contain a plaintext data key."
+    )
+
+
+def test_identity_keyring_as_generator_and_no_data_encryption_key():
+    test_multi_keyring = MultiKeyring(generator=IdentityKeyring())
+    with pytest.raises(GenerateKeyError) as exc_info:
+        test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_without_data_key())
+    assert exc_info.match("Unable to generate data encryption key.")
+
+
+def test_number_of_encrypted_data_keys_without_generator_with_children():
+    test_multi_keyring = get_multi_keyring_with_no_generator()
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    assert len(test.encrypted_data_keys) == len(test_multi_keyring.children)
+
+
+def test_number_of_encrypted_data_keys_without_children_with_generator():
+    test_multi_keyring = get_multi_keyring_with_no_children()
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    assert len(test.encrypted_data_keys) == 1
+
+
+def test_number_of_encrypted_data_keys_with_generator_and_children():
+    test_multi_keyring = get_multi_keyring_with_generator_and_children()
+    number_of_children = len(test_multi_keyring.children)
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    assert len(test.encrypted_data_keys) == number_of_children + 1
+
+
+def test_on_encrypt_when_data_encryption_key_given():
+    test_multi_keyring = MultiKeyring(generator=mock_generator, children=[mock_child_1, mock_child_2])
+    test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    for keyring in test_multi_keyring._decryption_keyrings:
+        keyring.on_encrypt.assert_called_once()
+
+
+def test_on_encrypt_edk_length_when_keyring_generates_but_does_not_encrypt():
+    test_multi_keyring = MultiKeyring(generator=OnlyGenerateKeyring())
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_without_data_key())
+    assert test.data_encryption_key is not None
+    assert len(test.encrypted_data_keys) == len(get_encryption_materials_without_data_key().encrypted_data_keys)
+
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_encrypted_data_key())
+    assert len(test.encrypted_data_keys) == len(get_encryption_materials_with_encrypted_data_key().encrypted_data_keys)
+
+
+def test_on_decrypt_when_data_encryption_key_given():
+    test_multi_keyring = MultiKeyring(generator=mock_generator, children=[mock_child_1, mock_child_2])
+    test_multi_keyring.on_decrypt(decryption_materials=get_decryption_materials_with_data_key(), encrypted_data_keys=[])
+    for keyring in test_multi_keyring._decryption_keyrings:
+        keyring.on_decrypt.assert_not_called()
```

Comment:
This is a problematic corner of Python's `mock` history...

Always use `assert not X.called`, not `X.assert_not_called()`.

The `assert_not_called` method was added in Python 3.5 because people kept assuming that it should be there; in earlier versions of Python 3 and in the `mock` library for Python 2, this will either silently work (because the `Mock.assert_not_called` is a mock and therefore returns a mock) or will throw an error, depending on the version of the `mock` library.

```suggestion
        assert not keyring.on_decrypt.called
```

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_not_called